README
see INSTALL for installation instructions


ChangeLog
2013-06-09  Tobias Schlemmer <keinstein@users.berlios.de>

	* configure.in, po/mutabor.pot: Revert "separate beta version string
	from version number by ~ to ensure order." This breaks OpenSuse
	Build service This reverts commit 2ece3d9ba9e6b4d1e6c65e82f53d8a7368ec6b21.
	* configure.in, po/mutabor.pot: separate beta version string from
	version number by ~ to ensure order.
	* osdep/linux/OpenSuse.spec.in: Add cppunit to OpenSuse spec.
	* debian/changelog, debian/control: fix some of the lintian
	complaints
	* debian/changelog: Update debian changelog and bump version.  Next try for compiling on lounchpad.
	* debian/control: Use libcppunit-dev for compiling on debian.
	* po/de.po, po/en.po, po/en_GB.po, po/mutabor.pot, po/ru.po: 
	Translation files and German translation.
	* configure.in: Check for Jack only if pkg-config is installed on
	linux.
	* tools/README, tools/fig/INSTALL, tools/fig/LICENSE,
	tools/fig/README, tools/fig/src/Makefile, tools/fig/src/control,
	tools/fig/src/control.c, tools/fig/src/control.h,
	tools/fig/src/control_parser.grm, tools/fig/src/execve.c,
	tools/fig/src/func.desc, tools/fig/src/launcher.c,
	tools/fig/src/log.c, tools/fig/src/log.h, tools/fig/src/loglevel.c,
	tools/fig/src/loglevel.h, tools/fig/src/make_control,
	tools/fig/src/malloc.c, tools/fig/src/prob.c, tools/fig/src/prob.h,
	tools/fig/src/reset, tools/fig/src/setup, tools/fig/src/spinlock.h,
	tools/fig/src/stubgen.awk, tools/fig/src/userfunc.h,
	tools/fig/src/util.c, tools/fig/src/util.h: add fig tool fig is a library to inject errors for testing purposes.

2013-06-08  Tobias Schlemmer <keinstein@users.berlios.de>

	* debian/control: add cppunit to Debian build depends
	* debian/changelog: Updated debian changelog

2013-06-09  Tobias Schlemmer <keinstein@users.berlios.de>

	* configure.in: Fix Jack detection on Mac OS X Assume no Jack is installed if pkg-config is not installed.  Don't
	stop, if Jack is not found.

2013-06-08  Tobias Schlemmer <keinstein@users.berlios.de>

	* src/kernel/Defs.h, src/kernel/Execute.cpp, src/kernel/Execute.h,
	src/kernel/GrafKern.cpp, src/kernel/Parser.cpp,
	src/kernel/TabGen.cpp, src/kernel/box.h, src/kernel/mut.cc,
	src/kernel/mut.h, src/kernel/mut.yy,
	src/kernel/routing/CommonFileDevice-inlines.h,
	src/kernel/routing/CommonFileDevice.h,
	src/kernel/routing/Device.cpp, src/kernel/routing/Device.h,
	src/kernel/routing/Route.h, src/kernel/routing/gmn/DevGIS.cpp,
	src/kernel/routing/gmn/DevGIS.h,
	src/kernel/routing/gmn/tests/gmnopenfail1.cpp,
	src/kernel/routing/gmn/tests/playgmn1.cpp,
	src/kernel/routing/midi/DevMidF.cpp,
	src/kernel/routing/midi/DevMidF.h,
	src/kernel/routing/midi/midicmn.cpp,
	src/kernel/routing/midi/midicmn.h,
	src/kernel/routing/midi/tests/DevMidFTest.cpp,
	src/kernel/routing/midi/tests/midicmnTest.h,
	src/kernel/routing/midi/tests/midiopenfail1.cpp,
	src/kernel/routing/tests/CommonFileDeviceTest.h,
	src/kernel/routing/tests/timer1.cpp, src/kernel/tests/boxTest.h,
	src/wxGUI/GUIBoxData.h, src/wxGUI/MutTextBox.cpp,
	src/wxGUI/Routing/DebugRoute.cpp,
	src/wxGUI/Routing/GUIRoute-inlines.h,
	src/wxGUI/generic/mutDebug.cpp: Fix compiler warnings.  A missing return statement caused strange behaviour of a test case.
	This and several similar cases have finally been found by gcc, and
	are corrected, now. Other warnings include comparisons of signed and
	unsigned integers, unused variables and wrong parenthesis.
	* configure.in: Reenable -Wall in debug mode.

2013-06-07  Tobias Schlemmer <keinstein@users.berlios.de>

	* m4/pkg.m4, m4/vector_data.m4: Add pkg.m4 and vector_data.m4 This adds tests for pkg-config and std::vector<...>::data()
	* common-rules.mk, configure.in: Enable building with jack support
	* src/kernel/routing/midi/tests/DevMidFTest.cpp,
	src/kernel/routing/midi/tests/DevMidFTest.h: Provide a sleep
	function for Windows in certain tests.

2013-06-07  Tobias Schlemmer <keinstein@users.berlios.de>

	* src/wxGUI/MutApp.h: Fix for build error due to missing
	OnAssertFailure in vtable.

2013-06-05  Tobias Schlemmer <keinstein@users.berlios.de>

	* configure.in, src/kernel/routing/midi/DevMidF.h: Add quirk for
	compilers that do not support vector::data().
	* lib/rtmidi/RtMidi.cpp: fix douplicate type in RtMidi.cpp
	* src/wxGUI/Routing/GUIRoute.h: Declare GUIRouteFactory so that the
	compiler consideres the right class in the right namespace.

2013-05-30  Tobias Schlemmer <keinstein@users.berlios.de>

	* templates/template.cpp: Add config to aviaillable groups in C++
	file template.
	* src/kernel/routing/Device.h, src/kernel/routing/midi/midicmn.cpp,
	src/kernel/routing/midi/tests/DevMidFTest.cpp: Prevent some
	controller redundancy from being sent.
	* src/wxGUI/MutRouteWnd.h, src/wxGUI/configtree.cpp,
	src/wxGUI/configtree.h: Remove some logs from file headers.
	* src/wxGUI/Routing/tests/main.cpp: Fix compiling.
	* src/kernel/routing/timing.h: fix time update in CurrentTimer.  The realtime condition was wrong.
	* src/kernel/Defs.h, src/kernel/routing/CommonFileDevice-inlines.h: 
	Fix building in non-debug mode
	* Makefile.am, tools/midi2text/Makefile.am: Add midi2text to
	compilation
	* src/kernel/routing/Device.h,
	src/kernel/routing/midi/tests/DevMidFTest.cpp: Fix shutdown of some
	tests.
	* src/kernel/routing/Device.h, src/kernel/routing/midi/midicmn.cpp,
	src/kernel/routing/midi/midicmn.h: Handle RPN and NRPN Controller
	values correctly.
	* src/kernel/routing/midi/DevMidF.cpp,
	src/kernel/routing/midi/DevMidF.h,
	src/kernel/routing/midi/midicmn-inlines.h,
	src/kernel/routing/tests/timing.h: Fix WriteDelta that created the
	wrong time.
	* src/kernel/routing/midi/DevMidF.h,
	src/kernel/routing/midi/tests/DevMidFTest.cpp,
	src/kernel/routing/midi/tests/DevMidFTest.h,
	src/kernel/routing/midi/tests/Makefile.am,
	src/kernel/routing/midi/tests/main.cpp,
	src/wxGUI/generic/mutDebug.cpp, src/wxGUI/generic/mutDebug.h,
	src/wxGUI/generic/mutDebugFlags.h: Add midi File recording test.

2013-05-29  Tobias Schlemmer <keinstein@users.berlios.de>


mu32/Defs.h
/** \file 
 ********************************************************************
 * Description
 *
 * $Header: /home/tobias/macbookbackup/Entwicklung/mutabor/cvs-backup/mutabor/mutabor/mu32/Defs.h,v 1.23 2011/10/04 17:16:13 keinstein Exp $
 * Copyright:   (c) 2008 TU Dresden
 * \author  Tobias Schlemmer <keinstein@users.berlios.de>
 * \date 
 * $Date: 2011/10/04 17:16:13 $
 * \version $Revision: 1.23 $
 * \license GPL
 *
 * $Log: Defs.h,v $
 * Revision 1.23  2011/10/04 17:16:13  keinstein
 * make program compile on Mac (wx 2.9) and fix some memory corruption
 *
 * Revision 1.22  2011-09-30 18:07:04  keinstein
 * * make compile on windows
 * * s/wxASSERT/mutASSERT/g to get assert handler completely removed
 * * add ax_boost_base for boost detection
 *
 * Revision 1.21  2011-09-29 05:26:58  keinstein
 * debug intrusive_ptr
 * fix storage and retrieving of input/output devices in treestorage
 * save maximum border size in icons
 * Apply the calculated offset in IconShape (box and box channels still missing)
 * Fix debug saving and restoring route information/route window on activation
 * Add wxWANTS_CHARS to MutEditWindow
 *
 * Revision 1.20  2011-09-28 07:35:53  keinstein
 * Make distclean happy
 *
 * Revision 1.19  2011-09-28 05:35:47  keinstein
 * fix compiling on ubuntu
 *
 * Revision 1.18  2011-09-27 20:13:20  keinstein
 * * Reworked route editing backend
 * * rewireing is done by RouteClass/GUIRoute now
 * * other classes forward most requests to this pair
 * * many bugfixes
 * * Version change: We are reaching beta phase now
 *
 * Revision 1.17  2011-09-06 08:09:20  keinstein
 * fix a compiler error showing a corruped error message
 *
 * Revision 1.16  2011-02-20 22:35:55  keinstein
 * updated license information; some file headers have to be revised, though
 *
 * Revision 1.2  2010-11-21 13:15:51  keinstein
 * merged experimental_tobias
 *
 * Revision 1.1.2.1  2010-01-11 10:12:59  keinstein
 * added some .cvsignore files
 *
 *
 ********************************************************************
 * \addtogroup kernel
 * \{
 ********************************************************************/
// prevent double inclusion

#ifndef	MU32_DEFS_H
#define MU32_DEFS_H

#ifdef HAVE_CONFIG_H
#include "config.h"
#else
#define MUTWIN
//#define MUTMIDI
#endif

#ifdef WX
#include "mhDefs.h"
#include "wx/wxchar.h"
#include "wx/intl.h"
#endif


#if defined(MUTWIN) && (!defined(WX) || defined(__WXMSW__))
#include <windows.h>
#include "wx/msw/winundef.h"
#endif

#ifdef WX
#include "stdint.h"
#  include "wx/setup.h"
#  include "wx/wxchar.h"
#  if !defined(__WXMSW__)
#define UINT unsigned int
#define WORD uint16_t
#define DWORD uint32_t
#define BYTE uint8_t
#define BOOL bool
#define pascal
#define CALLBACK
#endif
#define _export
#if defined(__VISUALC__)
#define min(a, b) wxMin(a, b)
#endif
#else // not WX
#define REUSE(type) type
#endif

#if defined(WX)
#define STD_PRE std
#else
#define STD_PRE
#endif

#define mutT _T

#ifdef WX
#define mut_thread_mutex(name)
#define mut_thread_locker(name)
#define mut_thread_lock(name) do {} while (0)
#define mut_thread_release(name) do {} while (0)


#define mutChar   wxChar
#define mutString wxString
#define mutStringRef mutString &
#define mutEmptyString (wxString) wxEmptyString
#define mutDelString(string) (string = mutEmptyString)
#define mutFreeString(string)
#define mutFopen  wxFopen
#define mutCopyString(left,right) (left = right)
#define mutCopyIntoString(left,right) (left = right)
#define mutStrdup wxStrdup
#define mutStrCmp(left, right) (left.Cmp (right))
#define mutStrEq(left, right) (left == right)
#define mutStrEq2(left, right) (left.IsSameAs(right))
#define mutStrLast(x) ((x).Last())
#define mutLen(x) ((x).Len())
#define C_STR(x) ((const wxChar *)(x))
#define mutC_STR(x) (C_STR(muT(x)))

#define mutStrLen wxStrlen
#define mutStrChr wxStrchr
#define mutIsspace wxIsspace
#define mutIsdigit wxIsdigit
#define mutFileName(name) (name.fn_str())

#define mutOFstream wxOutputStream
#define mutIFstream wxInputStream
#define mutTextStream wxTextFile

#define mutOpenOFstream(name,filename) \
	wxFFileOutputStream name (filename, _T("wb"))
#define mutOpenIFstream(name,filename) \
	wxFFileInputStream name (filename, _T("rb"))
#define mutOpenITextStream(name, filename) \
	wxTextFile name (filename)

#define mutWriteStream(stream,data,count) \
	(stream).Write(data,count)
#define mutReadStream(stream,data,count) \
	(stream).Read(data,count)
#define mutCloseStream(stream)

#define mutPutC(stream,data) (stream).PutC(data)
#define mutGetC(stream) (stream).GetC()

#define mutStreamBad(stream) (!(stream).IsOk())
#define mutStreamGood(stream) ((stream).IsOk())
#define mutStreamEOF(stream) ((stream).Eof())

#if wxCHECK_VERSION(2,9,0)
#define mutTranslationChar char
#define N_T(s) (s)
#define _C_STR(s) C_STR(s)
#else
#define mutTranslationChar mutChar
#define N_T(s) _T(s)
#define _C_STR(s) (s)
#endif

#else

#define mutChar char
#define mutString (char*)
#define mutStringRef mutString
#define mutEmptyString ((char *) NULL)
#define mutFreeString(string) if (string) free (string)
#define mutDelString(string) (mutFreeString(string), string = mutEmptyString)
#define mutFopen fopen
#define mutCopyString(left,right) left = strdup(right)
#define mutCopyIntoString(left,right) strcpy(left,right)
#define mutStrdup strdup
#define mutStrCmp(left,right) strcmp (left, right)
#define mutStrEq(left,right)  (!strcmp (left, right))
#define mutStrLast(x) (x[strlen(x)])
#define C_STR(x) (x)
#define mutC_STR(x) (x)


#define mutStrLen strlen
#define mutStrChr strchr
#define mutFileName

#define mutOFstream STD_PRE::ofstream
#define mutIFstream STD_PRE::ifstream
#define mutTextStrem STD_PRE::ifstream

#define mutOpenOFstream(name,filename) \
   STD_PRE::ofstream name(mutFileName(filename), STD_PRE::ios::out | STD_PRE::ios::binary/*0, filebuf::openprot*/)
#define mutOpenIFstream(name,filename) \
   STD_PRE::ifstream name(mutFileName(filename), STD_PRE::ios::in | STD_PRE::ios::binary/*0, filebuf::openprot*/)

#define mutWriteStream(stream,data,count) \
	stream.write(data,count)
#define mutReadStream(stream,data,count) \
	stream.read(data,count)
#define mutCloseStream(stream) stream.close()


#define mutPutC(stream,data) stream.putc(data)
#define mutGetC(stream) stream.getc()

#define mutStreamBad(stream) (stream.bad())
#define mutStreamGood(stream) (!stream.bad())
#define mutStreamEOF(stream) (stream.eof())

#endif

#ifdef DEBUG
#define mutAssertMsg(cond,msg) \
   if (!(cond)) \
	 std::cerr << (const char *)(wxString(msg).ToUTF8()) << std::endl; \
   assert(cond)
#define mutASSERT wxASSERT
#else
#define mutAssertMsg(cond,msg)
#define mutASSERT(cond)
#endif

inline wxString getContextLocal(const wxString & s)
{
	wxString ret = s.AfterFirst('|');

	if (ret == wxEmptyString) return s;
	else return ret;
}



class intrusive_ptr_refcount_type {
private:
	size_t value;

	template<class intrusive_ptr_T>
		friend void intrusive_ptr_add_ref(intrusive_ptr_T * obj);	
	template<class intrusive_ptr_T>
		friend void intrusive_ptr_release(intrusive_ptr_T * obj);
	template <class intrusive_ptr_T>
		friend size_t intrusive_ptr_get_refcount(intrusive_ptr_T * obj);
public:
	mut_thread_mutex(mutex)						
	intrusive_ptr_refcount_type():value(0) {}

	void lock () {
		mut_thread_lock(mutex);
	}

	void unlock () {
		mut_thread_release(mutex);
	}

	size_t operator ++() {
		lock();
		value++;
		unlock();
		return value;
	}

	size_t operator ++(int) {
		size_t retval = value;
		++(*this);
		return retval;
	}

	size_t operator --() {
		lock();
		value--;
		unlock();
		return value;
	}

	size_t operator --(int) {
		size_t retval = value;
		--(*this);
		return retval;
	}

	operator size_t () const {
		return value;
	}
};			

template <class intrusive_ptr_T>
inline size_t intrusive_ptr_get_refcount(intrusive_ptr_T * obj)
{
	if (!obj) return 0;
	return obj->intrusive_ptr_refcount;
}

template <class intrusive_ptr_T>
inline void intrusive_ptr_add_ref(intrusive_ptr_T * obj)
{
	if (!obj) return;
	++(obj->intrusive_ptr_refcount);
	DEBUGLOGTYPE(smartptr,*obj,_T("Incrementing %p to %d"),
		  obj,intrusive_ptr_get_refcount(obj));
}

template <class intrusive_ptr_T>
inline void intrusive_ptr_release(intrusive_ptr_T * obj)
{
	if (!obj) return;
	DEBUGLOGTYPE(smartptr,*obj,_T("Decrementing %p from %d"),
		  obj,intrusive_ptr_get_refcount(obj));
	if (!(--(obj->intrusive_ptr_refcount))) delete obj;
}

							      
#define REFPTR_INTERFACE						\
	public:								\
	intrusive_ptr_refcount_type intrusive_ptr_refcount;		


#define CHECK_REFPTR_NULL(class_data)				\
	mutASSERT(intrusive_ptr_get_refcount(class_data) <= 1);

#define MIDI_MIN_CHANNEL 0
#define MIDI_MAX_CHANNEL 15
#define MIDI_MIN_KEY 0
#define MIDI_MAX_KEY 0x7f
#endif /* MU32_DEFS_H */




///\}
